pkg/initializer/initializer.go

```diff
@@ -71,20 +71,33 @@ func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int6
 		if err != nil {
 			logger.Fatalf("failed to create clientset, %v", err)
 		}
+
 		m := member.NewMemberControl(e.Config.EtcdConnectionConfig)
-		isScaleup, err := m.IsClusterScaledUp(ctx, clientSet)
-		if err != nil {
-			logger.Errorf("scale-up not detected: %v", err)
-		} else if isScaleup {
-			logger.Info("Etcd cluster scale-up is detected")
-			// Add a learner(non-voting member) to a etcd cluster with retry
-			// If backup-restore is unable to add a learner in a cluster
-			// restart the `initialization` by exiting the backup-restore.
-			if err := member.AddLearnerWithRetry(ctx, m, addLearnerAttempts, e.Config.RestoreOptions.Config.DataDir); err != nil {
-				logger.Fatalf("unable to add a learner in a cluster: %v", err)
+
+		// check heartbeat of etcd member
+		if memberHearbeatPresent, err = m.WasMemberInCluster(ctx, clientSet); memberHearbeatPresent {
+			logger.Info("member found to be already a part of the cluster")
+			logger.Info("skipping the scale-up check")
+		} else {
+			if err != nil {
+				logger.Errorf("unable to check heartbeat of etcd member: %v", err)
+			}
+			logger.Info("member heartbeat is not present")
+			logger.Info("backup-restore will start the scale-up check")
+			isScaleup, err := m.IsClusterScaledUp(ctx, clientSet)
+			if err != nil {
+				logger.Errorf("scale-up not detected: %v", err)
+			} else if isScaleup {
+				logger.Info("Etcd cluster scale-up is detected")
+				// Add a learner(non-voting member) to a etcd cluster with retry
+				// If backup-restore is unable to add a learner in a cluster
+				// restart the `initialization` by exiting the backup-restore.
+				if err := member.AddLearnerWithRetry(ctx, m, addLearnerAttempts, e.Config.RestoreOptions.Config.DataDir); err != nil {
+					logger.Fatalf("unable to add a learner in a cluster: %v", err)
+				}
+				// return here after adding learner(non-voting member) as no restoration or validation required.
+				return nil
```

Comment:
> Generally, if a function returns an error in its function signature, then the best practice for consumers of that function is to check the error first before proceeding further.


If `memberHeartbeatPresent` is `false` or `err!=nil` , in both cases I want control to move to `else{ ... }` part as I don't want to change the `Initialization` flow when `memberHeartbeatPresent` is `false` or `err!=nil`.

```
if memberHeartbeatPresent, err = m.WasMemberInCluster(ctx, clientSet); err != nil {
	return fmt.Errorf("unable to check heartbeat of etcd member: %v", err)
}
```
the code you suggested is not correct as when `WasMemberInCluster()` returns `err!=nil`, it returns the `error`, hence it's changing the `Initialization` flow  by stopping the initialization, I don't want that.
`func WasMemberInCluster(){...}` shouldn't stop the initialization if it's return error as I don't want to change the `Initialization` flow.
If `memberHeartbeatPresent` is `false` or `err!=nil` , in both cases I want control to move to `else{ ... }` part and check for scale-up conditions

Reply:
> the code you suggested is not correct as when WasMemberInCluster() returns err!=nil, it returns the error, hence it's changing the Initialization flow by stopping the initialization, I don't want that

Sure, then you can still log the error and process. That will look something like
```code
if memberHeartbeatPresent, err = m.WasMemberInCluster(ctx, clientSet); err != nil {
    logger.Errorf("unable to check heartbeat of etcd member: %v", err)
} else if memberHeartbeatPresent {
	...
} else {
	...
}
```

I would still say that you should first check the error before you proceed to check the return value(s), because the return value cannot be trusted if `err!=nil`, because the called function/method is free to return any value as long as it returns the correct error, and it is the responsibility of the caller to ensure that it first checks the error before it checks the return values. Maybe this [blog post](https://earthly.dev/blog/golang-errors/#:~:text=When%20we%20do%20return%20an,returned%20are%20not%20relevant.) will help clarify this.

Reply:
> if memberHeartbeatPresent, err = m.WasMemberInCluster(ctx, clientSet); err != nil {
    logger.Errorf("unable to check heartbeat of etcd member: %v", err)
} else if memberHeartbeatPresent {
	...
} else {
	...
}

code is still wrong, you can't log the error and move to else part at the same time, hence this code will still changes the `Initialization` flow by stopping the initialization.
What I can do is change the singnature of func call with something like this.
```go
func WasMemberInCluster() bool{
// error occurs return false
// no error occurs return false/true accordingly
}
```
Please let me know if this seems fine to you.

Reply:
I have made the necessary changes, please let know if this doesn't seems fine to you.

Reply:
Thanks @ishan16696 , I understood your point now 👍 . One minor change though. Instead of

```
if memberHeartbeatPresent = m.WasMemberInCluster(ctx, clientSet); memberHeartbeatPresent {
...
}
```

you should rather do away with the variable `memberHeartbeatPresent` entirely, and use the simple condition

```
if m.WasMemberInCluster(ctx, clientSet) {
...
}
```

Reply:
actually we are using this `memberHeartbeatPresent ` boolean [here](https://github.com/gardener/etcd-backup-restore/pull/661/files#diff-dc0aa715ea97877db344484988dcdffda1fc0eaeda89257be6c70e9d724721d7R125).
Although I have never seen this condition getting true but it's still good as second line of defence and safety check.

Reply:
Ah ok, didn't see that. Thanks for the clarification.
/lgtm